(feat): refatorando o teste de backend para envolver mais lógica (#5)
Coderockr/backend-test: README.md

# Back End Test Project <img src="https://coderockr.com/assets/images/coderockr.svg" align="right" height="50px" />

You should see this challenge as an opportunity to create an application following modern development best practices (given the stack of your choice), but also feel free to use your own architecture preferences (coding standards, code organization, third-party libraries, etc). It’s perfectly fine to use vanilla code or any framework or libraries.

## Scope

In this challenge you should build an API for an application that stores and manages investments, it should have the following features:

  1. __Creation__ of an investment with a owner, a creation date and an amount.

>     1. An investment can be in the past.

Author: é referente a hoje na verdade, acho q talvez
```suggestion
    1. The creation date of a investment can be today or at the past.
```

    2. An investment should not be or became negative.

>   2. __View__ of a investment with its withdraws and current balance.

Author: `withdrawal` eh bem estranho

    1. Current balance should be the sum of the invested amount, the [interest] and withdraws.

>   3. Partial or total __withdraw__ of the investment amount.

Reviewer: ```suggestion
  3. Partial or total __withdrawal__ of the investment amount.
```

    1. Should not be able to withdraw more than the available balance.

>     2. Withdraws in the past can be created, but can't happen before the investment creation or previously created withdrawals.

Reviewer: ```suggestion
    2. Withdrawals in the past can be created, but can't happen before the investment creation or previously created withdrawals.
```

    3. Should show the amount paid on [taxes].

>   4. __List__ of all investments of a person.

Reviewer: List of a person's investments (com ou sem paginação?)

Author: eu ia deixar livre, mas acho válido colocar paginação sim.

__NOTE:__ the implementation of an interface will not be evaluated.

### Interest Calculation

> The investment will pay 0.52% every month in the same day of the investment creation, if a withdraw is made than in that month no interest will be paid.

Reviewer: then

> The investment will pay 0.52% every month in the same day of the investment creation, if a withdraw is made then in that month no interest will be paid.

Reviewer: ```suggestion
The investment will pay 0.52% every month in the same day of the investment creation, if a withdrawal is made then in that month no interest will be paid.
```

> Given that the interest is paid every month, than it should be treated as as compound interest, that means that interest that is not withdrawn will became part of the investment balance for the next interest payment.

Reviewer: pode remover o than (seria then mas não precisa dele ali)

Reviewer: which means
will become

### Taxation

> Every time a withdrawn happens part of it will automatically be kept as taxes.

Reviewer: withdrawal

> The taxes apply only on the interest, that means that if the investment was created with 1000.00, and when now the balance was 1,100.00, if you withdraw 110.00 than the taxes will be calculated over 11.00 (10% = 1,100/1000).

Reviewer: The taxes apply only to the interest, meaning that if ..., and now the balance is ..., ... then ...

Não ficou claro essa parte do 11.00

Author: a ideia eh dizer q a taxa se aplica apenas a uma parte do saque, tipo se na hora de sacar 10% do saldo (100) é de juros pago, então a taxa vai ser aplicada sobre 10% do q estiver sacado, no caso se sacar 110 então 11 representa os 10% de lucro, q a taxa deve ser sobre esses 11.

essa parte eu tinha achado dificil de escrever, e pelo visto ainda ficou ruim...

Reviewer: Sugestão de texto:

When money is withdrawn, tax is triggered. Taxes apply only to the interest portion of the money withdrawn. For example, if the initial investment was 1000.00, the current balance is 1200.00, and you withdraw 150.00, then the taxes will be applied to 16.67% = (200/1200) of the amount withdrawn. Meaning that you should apply tax on 25.00 (150 x 16,67%).

> The percent of taxes change with how old the investment is:

Reviewer: The tax percentage changes according to the age of the investment:

* If it is less than one year old, the percent will be 22.5%
* If it is between one and two years old, the percent will be 18.5%.

> * If older than two years, the percent will be 15%.

Reviewer: percentage

    1. The creation date of a investment can be today or an date in the past.
    2. An investment should not be or become negative.
  2. __View__ of an investment with its withdrawals and current balance.
    1. Current balance should be the sum of the invested amount, the [interest] and withdraws.
  3. Partial or total __withdraw__ of the investment amount.
    1. Should not be able to withdraw more than the available balance.
    2. Withdraws in the past can be created, but can't happen before the investment creation or previously created withdrawals.
    3. Should show the amount paid on [taxes].
  4. __List__ of a person's investments
    1. This list should have pagination.

__NOTE:__ the implementation of an interface will not be evaluated.

### Interest Calculation

The investment will pay 0.52% every month in the same day of the investment creation, if a withdraw is made then in that month no interest will be paid.

Given that the interest is paid every month, it should be treated as as compound interest, which means that interest that is not withdrawn will become part of the investment balance for the next interest payment.

### Taxation

> Every time a withdrawal happens part of it will automatically be kept as taxes.

Author: ```suggestion
```

> The taxes apply only to the interest, meaning that if the investment was created with 1000.00, and now the balance is 1,100.00, if you withdraw 110.00 than the taxes will be calculated over 11.00 (10% = 1,100/1000).

Author: ```suggestion
When money is withdrawn, tax is triggered. Taxes apply only to the interest portion of the money withdrawn. For example, if the initial investment was 1000.00, the current balance is 1200.00, and you withdraw 150.00, then the taxes will be applied to 16.67% = (200/1200) of the amount withdrawn. Meaning that you should apply tax on 25.00 (150 x 16,67%).
```
